pkg/services/store/entity/sqlstash/sql_storage_server.go

```diff
@@ -1106,12 +1182,337 @@ func (s *sqlEntityServer) List(ctx context.Context, r *entity.EntityListRequest)
+			key, err := entity.ParseKey(k)
+			if err != nil {
+				return err
+			}
+
+			args = append(args, key.Namespace, key.Group, key.Resource)
+			whereclause := "(" + s.dialect.Quote("namespace") + "=? AND " + s.dialect.Quote("group") + "=? AND " + s.dialect.Quote("resource") + "=?"
+			if key.Name != "" {
+				args = append(args, key.Name)
+				whereclause += " AND " + s.dialect.Quote("name") + "=?"
+			}
+			whereclause += ")"
+
+			where = append(where, whereclause)
+		}
+
+		entityQuery.addWhere("("+strings.Join(where, " OR ")+")", args...)
+	}
+
+	// Folder guid
+	if r.Folder != "" {
+		entityQuery.addWhere("folder", r.Folder)
+	}
+
+	if len(r.Labels) > 0 {
+		var args []any
+		var conditions []string
+		for labelKey, labelValue := range r.Labels {
+			args = append(args, labelKey)
+			args = append(args, labelValue)
+			conditions = append(conditions, "(label = ? AND value = ?)")
+		}
+		query := "SELECT guid FROM entity_labels" +
+			" WHERE (" + strings.Join(conditions, " OR ") + ")" +
+			" GROUP BY guid" +
+			" HAVING COUNT(label) = ?"
+		args = append(args, len(r.Labels))
+
+		entityQuery.addWhereInSubquery("guid", query, args)
+	}
+
+	entityQuery.addOrderBy("resource_version", Ascending)
+
+	var err error
+
+	s.log.Debug("watch init", "since", r.Since)
+
+	for hasmore := true; hasmore; {
+		err = func() error {
+			query, args := entityQuery.toQuery()
+
+			rows, err := s.sess.Query(ctx, query, args...)
+			if err != nil {
+				return err
+			}
+			defer func() { _ = rows.Close() }()
+
+			found := int64(0)
+
+			for rows.Next() {
+				found++
+				if found > entityQuery.limit {
+					entityQuery.offset += entityQuery.limit
+					return nil
+				}
+
+				result, err := s.rowToEntity(ctx, rows, rr)
+				if err != nil {
+					return err
+				}
+
+				if result.ResourceVersion > r.Since {
+					r.Since = result.ResourceVersion
+				}
+
+				if fromZero {
+					result.Action = entity.Entity_CREATED
+				}
+
+				s.log.Debug("sending init event", "guid", result.Guid, "action", result.Action, "rv", result.ResourceVersion)
+				err = w.Send(&entity.EntityWatchResponse{
+					Timestamp: time.Now().UnixMilli(),
+					Entity:    result,
+				})
+				if err != nil {
+					return err
+				}
+			}
+
+			hasmore = false
+			return nil
+		}()
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s *sqlEntityServer) poller(stream chan *entity.Entity) {
+	var err error
+	since := s.snowflake.Generate().Int64()
+
+	t := time.NewTicker(5 * time.Second)
+	defer t.Stop()
+
+	for range t.C {
+		since, err = s.poll(context.Background(), since, stream)
+		if err != nil {
+			s.log.Error("watch error", "err", err)
+		}
+	}
+}
+
+func (s *sqlEntityServer) poll(ctx context.Context, since int64, out chan *entity.Entity) (int64, error) {
+	s.log.Debug("watch poll", "since", since)
+
+	rr := &entity.ReadEntityRequest{
+		WithBody:   true,
+		WithStatus: true,
+	}
+
+	fields := s.getReadFields(rr)
+
+	entityQuery := selectQuery{
```

Comment:
Do we need [SELECT FOR SHARE](https://dev.mysql.com/doc/refman/8.0/en/innodb-locking-reads.html) ? This would ensure we don't skip entities in the process of being committed. I suspect transaction t1 with revision_version=1 might be comited after t2 with revision_version=2 leading to a gap? If that's correct, using  SELECT FOR SHARE would ensure we wait for all the in-progress transactions to finish before returning.

Reply:
Interesting, I'll have to think about that. It's supported by MySQL and Postgres but not SQLite. It does seem like it shouldn't hurt anything if we added it via dialect.